std/digest/hmac.d

```diff
@@ -0,0 +1,305 @@
+// Written in the D programming language.
+
+/**
+This package implements the hash-based message authentication code (_HMAC)
+algorithm as defined in $(WEB tools.ietf.org/html/rfc2104, RFC2104). See also
+the corresponding $(WEB en.wikipedia.org/wiki/Hash-based_message_authentication_code, Wikipedia article).
+
+$(SCRIPT inhibitQuickIndex = 1;)
+
+Macros:
+WIKI = Phobos/StdDigestHMAC
+SUBMODULE = $(LINK2 std_digest_$1.html, std.digest.$1)
+SUBREF = $(LINK2 std_digest_$1.html#.$2, $(TT $2))$(NBSP)
+
+License: $(WEB boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+Source: $(PHOBOSSRC std/digest/_hmac.d)
+ */
+
+module std.digest.hmac;
+
+import std.digest.digest : isDigest, hasBlockSize, isDigestibleRange, DigestType;
+import std.traits : isArray;
+import std.typetuple : allSatisfy;
+
+/**
+ * Template API HMAC implementation.
+ *
+ * This implements an _HMAC over the digest H. If H doesn't provide
+ * information about the block size, it can be supplied explicitly using
+ * the second overload.
+ *
+ * This type conforms to $(XREF digest.digest, isDigest).
+ */
+
+version(StdDdoc)
+/// Computes an HMAC over data read from stdin.
+unittest
+{
+    import std.stdio;
+    import std.digest.hmac, std.digest.sha;
+    import std.string : representation;
+
+    auto secret = "secret".representation;
+    stdin.byChunk(4096)
+         .hmac!SHA1(secret)
+         .toHexString!(LetterCase.lower)
+         .writeln;
+}
+
+template HMAC(H)
+if(isDigest!H && hasBlockSize!H)
+{
+    alias HMAC = HMAC!(H, H.blockSize);
+}
+
+/// ditto
+
+struct HMAC(H, size_t hashBlockSize)
+if(hashBlockSize % 8 == 0)
+{
+    enum blockSize = hashBlockSize;
+
+    private H digest;
+    private ubyte[blockSize / 8] key;
+
+    /**
+     * Constructs the HMAC digest using the specified secret.
+     */
+
+    this(scope const(ubyte)[] secret)
+    {
+        // if secret is too long, shorten it by computing its hash
+        typeof(digest.finish()) buffer = void;
+        if(secret.length > blockSize / 8)
+        {
+            digest.start();
+            digest.put(secret);
+            buffer = digest.finish();
+            secret = buffer[];
+        }
+
+        // if secret is too short, it will be padded with zeroes
+        // (the key buffer is already zero-initialized)
+        import std.algorithm.mutation : copy;
+        secret.copy(key[]);
+
+        start();
+    }
+
+    ///
+    @safe pure nothrow @nogc unittest
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        auto hmac = HMAC!SHA1("My s3cR3T keY".representation);
```

Comment:
This doesn't really test anything and pumps up coverage numbers giving a false sense of security. It should assert that e.g. the contentless digest is equal to whatever appropriate value.